src/prototype_room_external.js

```diff
@@ -277,33 +277,86 @@ Room.prototype.checkAndSpawnReserver = function() {
 Room.prototype.handleReservedRoom = function() {
   this.memory.state = 'Reserved';
   this.memory.lastSeen = Game.time;
-  if (this.memory.lastChecked !== undefined &&
-    Game.time - this.memory.lastChecked < 500) {
+
+  if (this.memory.lastChecked !== undefined && Game.time - this.memory.lastChecked < 10) {
     return false;
   }
   this.memory.lastChecked = Game.time;
 
-  let idiotCreeps = this.find(FIND_HOSTILE_CREEPS, {
+  this.checkHostiles();
+
+  this.checkSourcers();
+
+  this.checkReservers();
+
+  return false;
+};
+
+Room.prototype.checkHostiles = function() {
+  let hostiles = this.getEnemys();
+  if (hostiles.length > 0) {
+    this.memory.lastHostile = Game.time;
+  }
+  if (Game.time - this.memory.lastHostile < 600) {
+    roomName = this.name;
+    if (!this.memory.defender_last_called || Game.time - this.memory.defender_last_called > 200) {
+      if (hostiles.length > _.filter(Game.creeps, function(c) {
+          return c.memory.role === 'defender' && c.memory.routing.targetRoom === roomName;
+        })) {
+        Game.rooms[this.memory.reservation.base].memory.queue.push({
+          role: 'defender',
+          routing: {
+            targetRoom: roomName
+          },
+        });
+        this.memory.defender_last_called = Game.time;
+      }
+    }
+  }
+
+  let idiotCreeps = _.filter(hostiles, {
     filter: this.findAttackCreeps
   });
   for (let idiotCreep of idiotCreeps) {
     brain.increaseIdiot(idiotCreep.owner.username);
   }
+};
+
+Room.prototype.checkSourcers = function() {
```

Comment:
I think the logic for external rooms and `my` rooms is / should be the same.

https://github.com/TooAngel/screeps/blob/master/src/prototype_room_creepbuilder.js#L362